sec2_the_ruby_programming_language/2.8_hashes.rb

```diff
@@ -0,0 +1,36 @@
+#Hashes a.k.a.dictionary
+
+samples_hash ={'a'=> 1, 'b'=> 2, 'c'=> 3}
+my_details = {'name' => 'mashrur','favcolor'=> 'red'}
+p my_details['favcolor']  #"red"
+p samples_hash['b']       #2
+p samples_hash            #{"a"=>1, "b"=>2, "c"=>3}
+
+another_hash = {a: 1, b: 2, c: 3}
+p another_hash        #{:a=>1, :b=>2, :c=>3}
+p another_hash[:a]    #1
+p samples_hash.keys   #["a", "b", "c"]
+p samples_hash.values #[1, 2, 3]
+
+samples_hash.each do |key, value|
+  puts "The class for key is #{key.class} and the value is #{value.class}"
+end
+#The class for key is String and the value is Integer
+#The class for key is String and the value is Integer
+#The class for key is String and the value is Integer
+
+another_hash.each do |key,value|
+  puts "The class for key is #{key.class} and the value is #{value.class}"
+end
+# The class for key is Symbol and the value is Integer
```

Comment:
Symbol is not a changeable value, hence, it is often used as the identifier.
String is changeable series of characters or bytes that is often used as data.

Reply:
what do you mean by "changable value"?

Reply:
For example, when there are two variables that are declared as a string and a symbol respectively as shown below:

`str = "Hello"`
'str1 = :Hello

If I want to perform concatenation on `str`
`str + = " , World!"`
The output of the code will be `"Hello, World!"`

However, when I want to concatenate `", World!"`
An error will occur.

This is due to the reason that strings and symbols differ in how are they being stored.
Whenever a string is written, it creates a new object with a new place in memory, even if the string is textually identical to an existing string. However, since symbols are immutable, they always refer to the same object and the same place of memory.

Reply:
> This is due to the reason that strings and symbols differ in how are they being stored.
Whenever a string is written, it creates a new object with a new place in memory, even if the string is textually identical to an existing string. However, since symbols are immutable, they always refer to the same object and the same place of memory.

This is the key. Remember this. You will understand much later :)